community/kernel/src/main/java/org/neo4j/kernel/api/index/IndexPopulator.java

```diff
@@ -104,18 +104,15 @@ void add( IndexEntryUpdate<?> update )
      * If {@code populationCompletedSuccessfully} is {@code true} then it must mark this index
      * as {@link InternalIndexState#ONLINE} so that future invocations of its parent
      * {@link SchemaIndexProvider#getInitialState(long, IndexDescriptor)} also returns {@link InternalIndexState#ONLINE}.
-     */
-    void close( boolean populationCompletedSuccessfully ) throws IOException;
-
-    /**
-     * Called then a population failed. The failure string should be stored for future retrieval by
-     * {@link SchemaIndexProvider#getPopulationFailure(long)}. Called before {@link #close(boolean)}
-     * if there was a failure during population.
+     * The failure string must be stored for future retrieval by
+     * {@link SchemaIndexProvider#getPopulationFailure(long)} if {@code populationCompletedSuccessfully} is {@code
+     * false}.
      *
-     * @param failure the description of the failure.
-     * @throws IOException if marking failed.
+     * @param populationCompletedSuccessfully true if index should be marked as {@link InternalIndexState#ONLINE},
+     * otherwise false.
+     * @param failureMessage Description of population failure to store if populationCompletedSuccessfully is false.
      */
-    void markAsFailed( String failure ) throws IOException;
+    void close( boolean populationCompletedSuccessfully, String failureMessage ) throws IOException;
```

Comment:
That's how it was, and we thought it was confusing. Previously `markAsFailed` was always followed by `close( false )` and `close( true )` was run on its own. It was annoying to implement. Not proud of how it looks but according to me it feels better now.